### ISSUE TITLE
Download all the submissions of all my students.

### ISSUE THREAD

**Issue by fgallaire** [C0D3D3V/Moodle-Downloader-2#178]
## Avoid duplicates
* [x ] This enhancement request has not already been raised before

Hello, as a teacher I would like to download all the submissions of all my students (to note them !).

I think moodle-dl could do it easily, but is only working from a student point of view.

**Comment by C0D3D3V** [comment]
Yes the API allows it. It should also not be to hard to implement it. Anyone is welcome to do it.

I can implement it after I finished the rework I'm working on https://github.com/C0D3D3V/Moodle-Downloader-2/tree/pr_use_asyncio

I guess I can take a look on it end of this week.

**Comment by C0D3D3V** [comment]
Btw moodle allows to download all submissions of all students at once. I used the function already to note my students. The only improvement would be that moodle-dl can do it kind of automaticlly (and in a none ziped file). 

An problem I cann see is that the teacher can see submissions before the deadline, so the question is if not final submissions before deadline should also be downloaded.

**Comment by C0D3D3V** [comment]
See https://www.umass.edu/it/support/moodle/downloadallstudentsubmissionsamoodleassignment

There is also a documentation from Moodle but that was the first search result, and also looks well written  :D

**Comment by fgallaire** [comment]
> By the way, in Moodle you can download all the submissions of all the students at once. I have already used the feature to note my students. The only improvement would be that moodle-dl can somehow do it automatically (and in a non-ziped file).

You can download all the submissions of one assignment at one, not all the submissions of all the assignments.

One other problem : you can download the submissions of all the students (too much), or the submissions of all the students of a single group (too few), but not the submissions of all the students of multiple groups (what we really need).

**Comment by C0D3D3V** [comment]
Sounds like a good motivation :)

**Comment by C0D3D3V** [comment]
I did not do as much progress as I hoped I would. So it will take a bit longer till I can work on this... At this weekend I can not work on this project...

**Comment by C0D3D3V** [comment]
So two more questions...
Don't worry about the technical details even if I write a little about them below.


What do you think is the best folder structure?

Should a student's submissions go into a folder named similar to when you download it through Moodle? So "student full name_matriculation number". That would mean at least one extra API call per course to get the names and matriculation numbers. Alternatively, the name could be named based on the student's userid. (The moodle app does that call per assignment but I guess we can just ask for all enrolled students in the course, that list will be often much bigger, but should be faster than a call per assignment, I will test that...).
Should these folders then be located directly in the Assignmet folder or rather "Assignment/submissions/Student folder xy". 

Should the grades of the students be downloaded as well?

This could in principle also be made optional. As far as I can see the API only allows to retrieve one grade of a student at a time (I mean the feedback files/comments, the grades themselves can be retrieved all at once). (Whereas you can actually download a CSV file with all grades via the web interface, maybe I can find something for all at once in the API...). Well, according to the current status, it looks like you have to query the evaluation for each student individually and for each assignment. Do you think that is useful? I don't think so at the moment.

**Comment by fgallaire** [comment]
Hello,

Thanks a lot for the question.

It seems to me that, at least in a first time, the simpler/cleaner the better, so only the name of the student and no grade downloads.

Submission/CourseGroup/StudentName looks fine, but CourseGroup/StudentName/Submission could be better in some cases.

**Comment by C0D3D3V** [comment]
I was just working on this... 
I encountered a small problem, as soon as you use groupings in assignments the userids you get from the API seem to be completely wrong xD I have to find out if this is due to my old moodle version, or if the API is broken all the way up to the latest version. 
In the latter case, it will be goofy to download the submissions properly, then I would have to use the slow way again and download every single submission for every single assignment.

**Comment by C0D3D3V** [comment]
https://github.com/moodle/moodle/commit/a7e1f4ebc07d499f8ca843f6430a6d613f686572

I think that is the fix for the problem. So I will try to update my moodle, and see if it is fixed :D

**Comment by C0D3D3V** [comment]
Oh I see -.- I did not understand how group assignments work. Actually the userid is allways 0 no matter who in the group is submitting. So we can only use the group name, and not the student names (alternative would be to collect all student names of a group)

But I thought there where a option to split a course into multiple groups, and one group is allowed to do this assignment, but every group member for themselves....

**Comment by C0D3D3V** [comment]
Instead of naming all students in the group folder name, I could also create a file inside that group folder with all students listed that are in that group

**Comment by C0D3D3V** [comment]
I just checked what structure the "download all submissions" from moodle creates. It seems to not even mention the groups, it just downloads for every single student the same files for one group.

**Comment by C0D3D3V** [comment]
If you want to test it before the release... you can try it out, I implemented it in the main branch. Feel free to give feedback... maybe you want a different file structure.

**Comment by C0D3D3V** [comment]
released

**Comment by fgallaire** [comment]
Luckily, my university's Moodle version is 3.11 !

1) It works great, thanks again !

2) ISTM that the "all_submissions" folder is useless (and annoying)

3) Download all the submissions of all the students seems too much for me, what I would like is to download all the submissions of MY students. Could restrictions be added ?
- via configuration (whitelist/blacklist of groups)
- automatically (only the groups I'm member of)

**Comment by C0D3D3V** [comment]
Are you really member of all your groups? O.o

**Comment by C0D3D3V** [comment]
The all_submissions folder is just there to separate the assignment files from the submissions. What would you like instead?

**Comment by C0D3D3V** [comment]
How do you think the restrictions via configuration (whitelist/blacklist of groups) would look like? Do you know all your group ids, or group names?

**Comment by fgallaire** [comment]
> Are you really member of all your groups? O.o

Not sure of what groups are, but I think it's the way we are working here.

![moodle](https://user-images.githubusercontent.com/74269/226420885-1c4cdf33-9c29-45dc-adec-dd91c6bb833c.jpg)

**Comment by fgallaire** [comment]
After checking this is not the case for all the courses, but it could be done easily

**Comment by C0D3D3V** [comment]
Please open a new issue with your change request / feature request. Please word it well, so that I can implement what you want.

**Comment by fgallaire** [comment]
> The all_submissions folder is just there to separate the assignment files from the submissions. What would you like instead?

Not sure of what assignments are exactly, and where they should be, but here are my folders, just would like the "all_submissions":

![moodle2](https://user-images.githubusercontent.com/74269/226423489-fdb39454-1531-4125-a37f-4b8e995719fa.jpg)

**Comment by C0D3D3V** [comment]
you probably do not use assignmet files.
Most courses I saw, will add Pdf files to an assignment with questions and tasks the student nedds to do. Also you probalby disabled downloading descriptions, because every assignment normaly has an introduction that will be in that folder too.

But please do not use this closed issue. Please open a new issue so I can track the progress there. For me is everything fine you want to request, ...

**Comment by C0D3D3V** [comment]
How do you want to configure the whitelist?
